<a href="https://colab.research.google.com/github/ieee-saocarlos/2021EstudosPython/blob/main/2022/CS_dia3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Introdução ao Pandas</center>
___

## Conteúdo
3. [Pandas](#pandas)<br>

<a name="pandas"></a>
## 3. Pandas

### 3.1. O que é o Pandas?

Pandas é uma biblioteca *open source*, que proporciona estruturas de dados e ferramentas de análise de dados de alta performance e fáceis de usar para Python. Vamos entender melhor o que alguns termos significam:
 - *open source*: assim como o NumPy, seu código original é disponibilizado livremente e pode ser distribuido e modificado. Isso significa que qualquer um pode contribuir para a evolução do Pandas!
 - alta performance: Pandas é escrito em Python, Cython e C. Isso permite que os cientistas de dados consigam o utilizar para lidar com conjuntos de dados muito grandes (daqueles que o Excel não conseguiria nem abrir) e fazer operações sobre esses dados com facilidade. Dessa forma, Pandas torna nosso trabalho melhor e mais fácil provendo ótima performânce.
 - estruturas de dados e análises de dados: o motivo pelo qual Pandas existe. Muitas vezes precisamos obter insights de dados crus como documentos de textos, tabelas e etc. Pandas é capaz de lidar com esses tipos de dados para que possamos analisá-los.

Em resumo, **Pandas fornece estrutura de dados especializadas e ferramentas para manipulação de dados**. Sua ótima performance, facilidade de uso e comunidade dedicada são as principais razões de sua vasta adoção entre cientistas de dados. 

Agora vamos começar a utilizá-lo! Se você está utilizando o Anaconda ou o Google Colab, já deve ter ele instalado. Senão, é possível achar os passos de instalação [neste link](https://pandas.pydata.org/). <br>

### 3.2. Básico de Pandas

Vamos relembrar as etapas da construção de um modelo de dados para entender a importância do Pandas:

1. Entendimento do problema / Definição do escopo
2. Definição do objetivo e métricas de avaliação
3. Determinação dos dados necessários
4. Aquisição dos dados
5. Tratamento e manipulação dos dados
6. Análise Exploratória dos Dados (E.D.A.)
7. Engenharia de Variáveis
8. Construção e avaliação do modelo
9. Interpretação dos resultados e apresentação dos dados
10. Deploy
11. Monitoramento e manutenção

Então, mas em quais etapas o Pandas é mais utilizado? Nas etapas 4 a 7 e 9! Ou seja, das 11 etapas, o Pandas pode nos ajudar em 5 delas! Isso mostra o quão poderosa essa biblioteca é, e porque utilizamos ela diariamente.

#### 3.2.1. Como importá-lo?

O Pandas é geralmente importado sempre da mesma forma, como mostramos abaixo. Aproveitamos também para importar o NumPy, biblioteca que já apresentamos anteriormente.

In [ ]:
import pandas as pd
import numpy as np

#### 3.2.2. Objetos do Pandas

Existem 2 principais tipos de objetos no Pandas: as [*Series*](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series) e os *DataFrames*. As *Series* são sequências de uma dimensão  de elementos (pra ser mais específico *ndarray*), todos do mesmo tipo de dados, com rótulos/índices (*labels*). São o objeto primário do Pandas, tudo vai funcionar baseado nelas. Pra criar um objeto do tipo *Series*, podemos fazer o seguinte:

In [ ]:
s = pd.Series([1, 3, 5, np.nan, 6, 8])
s

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

Podemos ver acima que cada elemento da *Series* tem um rótulo relacionado. Esses rótulos podem ser tanto numéricos quanto de texto! Ao final do objeto temos a informação sobre o tipo de dados da *Series*: nesse caso, números *float64*. <br>
O outro objeto principal do Pandas é o [*DataFrame*](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html#pandas.DataFrame), que é basicamente uma coleção de *Series* com rótulos em comum. O *DataFrame* é bem parecido com uma tabela de Excel, com seus índices e colunas. Tanto suas linhas quanto suas colunas tem rótulos, nos permitindo acessar qualquer célula pela sua coordenada. Há diversas formas de criar um *DataFrame*, vamos começar com o mais simples:

In [ ]:
dates = pd.date_range('20210101', periods=6) #estamos criando uma lista de datas entre 01/01/2021 e 06/01/2021. Note que estamos usando uma função do Pandas
df = pd.DataFrame(
    np.random.randn(6, 4), #apenas números aleatórios nas células
    index=dates,  #especificando quais são os índices. Eles aceitam até datas como índice! Isso é muito bom para lidar com dados de séries temporais
    columns=list('ABCD')) #especificando como quero que sejam os nomes das colunas passando uma lista de letras
df

A         B         C         D
2021-01-01  2.465460  1.005221  1.323598  0.340436
2021-01-02 -0.750289  0.301581 -0.652988  0.593393
2021-01-03 -0.089641 -0.610242  0.598481  0.370903
2021-01-04  0.398810 -1.553834  0.208626  0.150176
2021-01-05 -0.036295 -0.901358 -0.502967 -1.704102
2021-01-06 -0.644397  1.253166 -1.993942 -1.646424

Isso é um *DataFrame*! Uma coisa boa do Pandas com o Jupyter Notebook é que eles mostram o *DataFrame* de uma forma bastante amigável. Agora que sabemos sobre as duas principais estruturas de dados do Pandas, podemos aprender sobre os principais métodos e funcionalidades dessa biblioteca, e para isso vamos utilizar dados reais sobre uma das marcas mais conhecidas dos desenhos e videogames, a franquia Pokémon :)

Observação: um *DataFrame* pode ser visto como um dicionário de listas.

#### 3.2.3. Carregando o conjunto de dados

Um dos tipos de dados mais comuns para se guardar arquivos são os CSVs (comma-separated values, em português, valores separados por vírgulas). O Pandas tem diversas funções para transformar os mais variados tipos de arquivos em *DataFrames*, como csv, Excel, json e etc. Nesse exemplo, vamos usar o [leitor de csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html). Execute a célula abaixo para carregar os dados dos Pokémons.

Caso você não conheça, [Pokémon](https://www.pokemon.com/us/) é uma franquia japonesa composta por jogos de videogame, desenhos, cartas colecionáveis, entre outros produtos. Nos jogos, principal produto da franquia, os humanos (conhecidos como treinadores Pokémon) capturam e treinam os monstrinhos (Pokémons) para realizar diversas aventuras e batalhas. Cada Pokémon pode ser de diferentes tipos (fogo, água, elétrico, pedra, etc) e apresentam diferentes níveis de habilidades (ataque, defesa, velocidade, etc). Além disso, alguns Pokémons são mais raros que outros (conhecidos como lendários).

Caso você esteja utilizando o Google Colab, não se esqueça de conectar no seu Drive e alterar o caminho para a pasta que contém seus arquivos de dados.

In [ ]:
pkmn = pd.read_csv(
    './pokemon_data.csv', #o caminho para o arquivo que se quer ler
    sep=',') #o caracter utilizado para separar os valores

#### 3.2.4. Visualizações iniciais
Ótimo! Acabamos de criar um *DataFrame* a partir de um arquivo csv. Mas como gostamos de verificar as coisas, seria interessante saber algumas coisas como: o que os dados contêm, como são as colunas, se tem valores nulos e etc. O Pandas tem 4 métodos principais para isso:

In [ ]:
pkmn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


A primeira coisa que é interessante de se fazer após ler um conjunto de dados em um *DataFrame* é utilizar o método *info()*. Ele mostra informações como:
 - a classe do objeto criado
 - o intervalo do índice e quantas linhas de dados se tem
 - as colunas, seus nomes e tipos de dados
 - quais os tipos de dados presentes no *DataFrame* e quantas colunas de cada
 - a quantidade de memória utilizada pelo computador para guardar esses dados

Podemos ver que temos uma base sobre Pokémon, com 800 linhas com índices númericos de 0 a 799, 13 colunas de três tipos de dados diferentes, usando cerca de ~76kB de memória RAM. As informações presentes sobre os Pokémons são o número, o nome, o tipo (alguns tem subtipo, mas não todos, por isso os dados faltantes na coluna *Type 2*), estatísticas de ataque, defesa e velocidade, a geração (novos Pokémons são apresentados a cada alguns anos, e cada grupo é conhecido por uma geração) e a indicação se ele é lendário (como uma espécie mística). <br> Todas essas informações a gente conseguiu descobrir com apenas uma linha de código! Vamos então ver como o *DataFrame* realmente é. Temos dois métodos para isso:

In [ ]:
pkmn.head()

#                   Name Type 1  ... Speed  Generation  Legendary
0  1              Bulbasaur  Grass  ...    45           1      False
1  2                Ivysaur  Grass  ...    60           1      False
2  3               Venusaur  Grass  ...    80           1      False
3  3  VenusaurMega Venusaur  Grass  ...    80           1      False
4  4             Charmander   Fire  ...    65           1      False

[5 rows x 13 columns]

In [ ]:
pkmn.tail()

#                 Name   Type 1  ... Speed  Generation  Legendary
795  719              Diancie     Rock  ...    50           6       True
796  719  DiancieMega Diancie     Rock  ...   110           6       True
797  720  HoopaHoopa Confined  Psychic  ...    70           6       True
798  720   HoopaHoopa Unbound  Psychic  ...    80           6       True
799  721            Volcanion     Fire  ...    70           6       True

[5 rows x 13 columns]

Os métodos *.head()* e *.tail()* mostram, respectivamente, as primeiras e últimas *n* linhas do *DataFrame* (por padrão, *n*=5, mas você pode passar qualquer número como parâmetro) mostrando os índices e os nomes das colunas como numa tabela.

In [ ]:
pkmn.describe()

#      Total          HP  ...     Sp. Def       Speed  Generation
count  800.000000  800.00000  800.000000  ...  800.000000  800.000000   800.00000
mean   362.813750  435.10250   69.258750  ...   71.902500   68.277500     3.32375
std    208.343798  119.96304   25.534669  ...   27.828916   29.060474     1.66129
min      1.000000  180.00000    1.000000  ...   20.000000    5.000000     1.00000
25%    184.750000  330.00000   50.000000  ...   50.000000   45.000000     2.00000
50%    364.500000  450.00000   65.000000  ...   70.000000   65.000000     3.00000
75%    539.250000  515.00000   80.000000  ...   90.000000   90.000000     5.00000
max    721.000000  780.00000  255.000000  ...  230.000000  180.000000     6.00000

[8 rows x 9 columns]

Por fim, o método *.describe()* mostra um resumo estatístico de todas as colunas numéricas. É um método bom para ter uma ideia inicial sobre o que ocorre nas colunas de uma perspectiva estatística.

### 3.3. Filtrando (*Filtering*) e fatiando (*slicing*) os dados

*Filtering* e *slicing* são técnicas utilizadas para isolar partes específicas do *DataFrame*, sejam linhas, colunas ou células. Isso é muito útil pois diversas vezes queremos analisar alguns dados ao invés da base inteira. O Pandas tem ferramentas próprias para isso. <br>

#### 3.3.1. *Slicing*

No Pandas existem duas principais formas de fatiar os dados, isto é, selecionar apenas uma parte de acordo com as linhas e colunas do *DataFrame*: utilizando o nome das partes ou com os métodos *.loc()* e *.iloc()*. Vamos começar pelo primeiro. Para fazer isso, imagine que queremos apenas os nomes e o poder de ataque dos Pokémons e veja o exemplo:

In [ ]:
pkmn[['Name','Attack']].head(7)

Name  Attack
0              Bulbasaur      49
1                Ivysaur      62
2               Venusaur      82
3  VenusaurMega Venusaur     100
4             Charmander      52
5             Charmeleon      64
6              Charizard      84

Duas coisas importantes aqui:
 - Juntamos o *slicing* com o método *.head()* na mesma linha, para que fosse possível ver o resultado do fatiamento. Ao usar o Pandas é possível e comum fazer esse tipo de agrupamento de operações.
 - Foram utilizadas chaves duplas ```[[]]``` no fatiamento. Ao fazer isso, estou explicitando que quero um objeto do tipo *DataFrame*. Se eu quisesse objetos do tipo *Series* usaria chaves simples. Podemos ver um exemplo disso abaixo:

In [ ]:
print(type(pkmn[['Name']]))
pkmn[['Name']]

<class 'pandas.core.frame.DataFrame'>


Name
0                Bulbasaur
1                  Ivysaur
2                 Venusaur
3    VenusaurMega Venusaur
4               Charmander
..                     ...
795                Diancie
796    DiancieMega Diancie
797    HoopaHoopa Confined
798     HoopaHoopa Unbound
799              Volcanion

[800 rows x 1 columns]

Outra forma de se obter um objeto do tipo *Series* é passando a coluna como se fosse um atributo do *DataFrame*:

In [ ]:
pkmn.Name

0                  Bulbasaur
1                    Ivysaur
2                   Venusaur
3      VenusaurMega Venusaur
4                 Charmander
               ...          
795                  Diancie
796      DiancieMega Diancie
797      HoopaHoopa Confined
798       HoopaHoopa Unbound
799                Volcanion
Name: Name, Length: 800, dtype: object

O único problema desse formato é que colunas cujo nome contém espaços não funcionarão, como é o caso das colunas *Type 1*, *Type 2*, *Sp. Atk* e *Sp. Def*. Para resolver isso, vamos renomeá-las com o método *.rename()*.

In [ ]:
pkmn.rename(
    columns={'Type 1':'Type_1', 'Type 2':'Type_2', 'Sp. Atk':'Sp_Atk','Sp. Def':'Sp_Def'}, #passando o nome antigo e novo como um dicionário
    inplace = True #algumas operações com Pandas criam uma cópia do DataFrame e não alteram o objeto em si, alteramos isso mudando o parâmetro inplace para verdadeiro
)

Agora que os nomes foram trocados, podemos obter um objeto *Series* do tipo do Pokémon como a seguir:

In [ ]:
pkmn.Type_1

0        Grass
1        Grass
2        Grass
3        Grass
4         Fire
        ...   
795       Rock
796       Rock
797    Psychic
798    Psychic
799       Fire
Name: Type_1, Length: 800, dtype: object

Outra forma de selecionar partes dos dados é usando os métodos *.loc()* e *.iloc()*.<br>
Para usar a localização númerica utilizamos o *iloc*. Como você pode imaginar, a linhas e colunas são ordenadas por números inteiros sequenciais, começando do 0, como nas listas. Dessa forma, se você sabe o número da linha e da coluna, você pode usar o *iloc*. Por exemplo, se quisermos a coluna HP, que é a 6ª, poderíamos fazer o seguinte:

In [ ]:
pkmn.iloc[:,5].head()

0    45
1    60
2    80
3    80
4    39
Name: HP, dtype: int64

A sintaxe do *iloc* é como [x,y], que significa que queremos a (x+1)ª linha e (y+1)ª coluna. Se utilizarmos ```:``` no lugar de x ou y significa que queremos a coluna ou linha completa, respectivamente. Vamos pegar o HP do Bulbasaur, o primeiro Pokémon do nosso *DataFrame*:

In [ ]:
print("O HP do Bulbasaur é "+str(pkmn.iloc[0,5]))

O HP do Bulbasaur é 45


O método *.loc()* usa o rótulo para acessar os valores. Dessa forma, ao invés de passarmos as coordenadas numéricas, passamos o nome da linha e da coluna, como a seguir:

In [ ]:
pkmn.iloc[0, :]

#                     1
Name          Bulbasaur
Type_1            Grass
Type_2           Poison
Total               318
HP                   45
Attack               49
Defense              49
Sp_Atk               65
Sp_Def               65
Speed                45
Generation            1
Legendary         False
Name: 0, dtype: object

In [ ]:
pkmn.loc[0, :]

#                     1
Name          Bulbasaur
Type_1            Grass
Type_2           Poison
Total               318
HP                   45
Attack               49
Defense              49
Sp_Atk               65
Sp_Def               65
Speed                45
Generation            1
Legendary         False
Name: 0, dtype: object

No caso, os rótulos das linhas são iguais às suas coordenadas, por isso ficou parecido com o *iloc*. Vamos fazer o teste com as colunas também para ver a diferença. Abaixo pegaremos novamente o HP do Bulbasaur:

In [ ]:
print("O HP do Bulbasaur é "+str(pkmn.loc[0,'HP']))

O HP do Bulbasaur é 45


#### 3.3.2. Filtros (*Filtering*)

Uma vez sabendo isolar partes do *DataFrame* de acordo com a localização dos dados, podemos partir para isolar de acordo com condições, ou seja, filtrar os dados.
Para conseguir fazer isso no Pandas, fazemos o seguinte: passamos uma expressão condicional e o Pandas retorna apenas as partes que teriam a condição como verdade. Para testar isso, vamos ver a defesa média de todos os Pokémons e depois ver se os tipos 'Rock' e 'Steel' têm defesas maiores:

In [ ]:
pkmn.Defense.mean() #note que operações comuns como média (mean), mediana (median) e soma (sum) são métodos do Pandas

73.8425

In [ ]:
pkmn.loc[pkmn.Type_1=='Fairy'].Defense.mean()

65.70588235294117

In [ ]:
pkmn.loc[pkmn.Type_1=='Dragon'].Defense.mean()

86.375

In [ ]:
pkmn.loc[pkmn.Type_1=='Rock'].Defense.mean()

100.79545454545455

In [ ]:
pkmn[pkmn.Type_1=='Steel'].Defense.mean()

126.37037037037037

De fato, parece que os tipos selecionados tem média acima dos demais Pokémons. Para verificar isso, passamos a condição pkmn.Type_1=='Steel' entre chaves, o que retorna apenas as linhas de tal tipo. Com isso, selecionamos apenas a coluna de defesa e calculamos a média. <br>
Vamos ver agora os Pokémons com defesa maior que 150, cujo tipo principal não é 'Rock' nem 'Steel':

In [ ]:
pkmn[(pkmn.Defense > 150)&(pkmn.Type_1!='Rock')&(pkmn.Type_1!='Steel')]

#                   Name   Type_1  ... Speed  Generation  Legendary
87    80    SlowbroMega Slowbro    Water  ...    30           1      False
98    91               Cloyster    Water  ...    70           1      False
230  213                Shuckle      Bug  ...     5           2      False
424  383  GroudonPrimal Groudon   Ground  ...    90           3       True
430  386    DeoxysDefense Forme  Psychic  ...    90           3       True
789  713                Avalugg      Ice  ...    28           6      False

[6 rows x 13 columns]

Podemos ver acima que é possível juntar condições com os operadores E (&) e OU (|), como vimos no inicio da aula.<br>
Vamos dizer agora que você quer apenas alguns Pokémons em específico, por exemplo Venusaur, Charizard e Blastoise. Criar uma condição para cada e uní-las com o operador & pode ser difícil, ainda mais se for uma quantidade grande de opções. Podemos facilitar isso passando uma tupla ao método *.isin()*, como abaixo:

In [ ]:
aux = ('Venusaur', 'Charizard', 'Blastoise')
pkmn[pkmn.Name.isin(aux)]

#       Name Type_1  Type_2  ...  Sp_Def  Speed  Generation  Legendary
2   3   Venusaur  Grass  Poison  ...     100     80           1      False
6   6  Charizard   Fire  Flying  ...      85    100           1      False
11  9  Blastoise  Water     NaN  ...     105     78           1      False

[3 rows x 13 columns]

Finalmente, podemos criar novos *DataFrames* a partir de um já existente selecionando apenas algumas linhas ou colunas dele:

In [ ]:
offensive_stats = pkmn[['#','Name','Attack','Sp_Atk','Speed']] #selecionando apenas estatísticas ofensivas
defensive_stats = pkmn[['#','Name', 'HP','Defense','Sp_Def']] #selecionando apenas estatísticas defensivas

In [ ]:
offensive_stats.head()

#                   Name  Attack  Sp_Atk  Speed
0  1              Bulbasaur      49      65     45
1  2                Ivysaur      62      80     60
2  3               Venusaur      82     100     80
3  3  VenusaurMega Venusaur     100     122     80
4  4             Charmander      52      60     65

In [ ]:
defensive_stats.head()

#                   Name  HP  Defense  Sp_Def
0  1              Bulbasaur  45       49      65
1  2                Ivysaur  60       63      80
2  3               Venusaur  80       83     100
3  3  VenusaurMega Venusaur  80      123     120
4  4             Charmander  39       43      50

In [ ]:
fire_pkmn = pkmn[(pkmn.Type_1=='Fire')|(pkmn.Type_2=='Fire')] #filtrando apenas linhas com algumas condições
fire_pkmn.head()

#                       Name Type_1  ... Speed  Generation  Legendary
4  4                 Charmander   Fire  ...    65           1      False
5  5                 Charmeleon   Fire  ...    80           1      False
6  6                  Charizard   Fire  ...   100           1      False
7  6  CharizardMega Charizard X   Fire  ...   100           1      False
8  6  CharizardMega Charizard Y   Fire  ...   100           1      False

[5 rows x 13 columns]

In [ ]:
water_pkmn = pkmn[(pkmn.Type_1=='Water')|(pkmn.Type_2=='Water')] #filtrando apenas linhas com algumas condições
water_pkmn.head()

#                     Name Type_1  ... Speed  Generation  Legendary
9    7                 Squirtle  Water  ...    43           1      False
10   8                Wartortle  Water  ...    58           1      False
11   9                Blastoise  Water  ...    78           1      False
12   9  BlastoiseMega Blastoise  Water  ...    78           1      False
59  54                  Psyduck  Water  ...    55           1      False

[5 rows x 13 columns]

### 3.4. Juntando DataFrames

É muito comum ter a necessidade de juntar *DataFrames* diferentes. Se você já utilizou SQL ou qualquer outro banco de dados relacional, deve conhecer isso como *join*. O Pandas também tem a mesma função utilizando o método *.merge()*. Antes do exemplo, vamos aprender/relembrar os tipos de *joins* mais comuns:<br>
![Joining Methods](https://i.imgur.com/HaSBT91.jpg) <br>
Agora, vamos testar os *merge* nos *DataFrames* filtrados que criamos anteriormente:

In [ ]:
all_stats = pd.merge(
    offensive_stats, #o DataDrame da esquerda
    defensive_stats, #o DataDrame da direita
    how='inner', #o tipo de join que queremos fazer
    on=['#','Name']) #baseado em quais valores em comum
all_stats.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 800 entries, 0 to 799
Data columns (total 8 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   #        800 non-null    int64 
 1   Name     800 non-null    object
 2   Attack   800 non-null    int64 
 3   Sp_Atk   800 non-null    int64 
 4   Speed    800 non-null    int64 
 5   HP       800 non-null    int64 
 6   Defense  800 non-null    int64 
 7   Sp_Def   800 non-null    int64 
dtypes: int64(7), object(1)
memory usage: 56.2+ KB


Ótimo! Conseguimos fazer o *merge* (termo mais utilizado no Pandas) de dois *DataDrames*. Lembre-se que *inner*, *left*, *right* e *outer* terão resultados diferentes:

In [ ]:
fire_pkmn.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64 entries, 4 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           64 non-null     int64 
 1   Name        64 non-null     object
 2   Type_1      64 non-null     object
 3   Type_2      36 non-null     object
 4   Total       64 non-null     int64 
 5   HP          64 non-null     int64 
 6   Attack      64 non-null     int64 
 7   Defense     64 non-null     int64 
 8   Sp_Atk      64 non-null     int64 
 9   Sp_Def      64 non-null     int64 
 10  Speed       64 non-null     int64 
 11  Generation  64 non-null     int64 
 12  Legendary   64 non-null     bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 6.6+ KB


In [ ]:
water_pkmn.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126 entries, 9 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           126 non-null    int64 
 1   Name        126 non-null    object
 2   Type_1      126 non-null    object
 3   Type_2      67 non-null     object
 4   Total       126 non-null    int64 
 5   HP          126 non-null    int64 
 6   Attack      126 non-null    int64 
 7   Defense     126 non-null    int64 
 8   Sp_Atk      126 non-null    int64 
 9   Sp_Def      126 non-null    int64 
 10  Speed       126 non-null    int64 
 11  Generation  126 non-null    int64 
 12  Legendary   126 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 12.9+ KB


In [ ]:
left_fire_water = pd.merge(fire_pkmn, water_pkmn, how='left', on=['#','Name'])
left_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 64 entries, 0 to 63
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   #             64 non-null     int64  
 1   Name          64 non-null     object 
 2   Type_1_x      64 non-null     object 
 3   Type_2_x      36 non-null     object 
 4   Total_x       64 non-null     int64  
 5   HP_x          64 non-null     int64  
 6   Attack_x      64 non-null     int64  
 7   Defense_x     64 non-null     int64  
 8   Sp_Atk_x      64 non-null     int64  
 9   Sp_Def_x      64 non-null     int64  
 10  Speed_x       64 non-null     int64  
 11  Generation_x  64 non-null     int64  
 12  Legendary_x   64 non-null     bool   
 13  Type_1_y      1 non-null      object 
 14  Type_2_y      1 non-null      object 
 15  Total_y       1 non-null      float64
 16  HP_y          1 non-null      float64
 17  Attack_y      1 non-null      float64
 18  Defense_y     1 non-null      fl

In [ ]:
right_fire_water = pd.merge(fire_pkmn, water_pkmn, how='right', on=['#','Name'])
right_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 126 entries, 0 to 125
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   #             126 non-null    int64  
 1   Name          126 non-null    object 
 2   Type_1_x      1 non-null      object 
 3   Type_2_x      1 non-null      object 
 4   Total_x       1 non-null      float64
 5   HP_x          1 non-null      float64
 6   Attack_x      1 non-null      float64
 7   Defense_x     1 non-null      float64
 8   Sp_Atk_x      1 non-null      float64
 9   Sp_Def_x      1 non-null      float64
 10  Speed_x       1 non-null      float64
 11  Generation_x  1 non-null      float64
 12  Legendary_x   1 non-null      object 
 13  Type_1_y      126 non-null    object 
 14  Type_2_y      67 non-null     object 
 15  Total_y       126 non-null    int64  
 16  HP_y          126 non-null    int64  
 17  Attack_y      126 non-null    int64  
 18  Defense_y     126 non-null    

In [ ]:
inner_fire_water = pd.merge(fire_pkmn, water_pkmn, how='inner', on=['#','Name'])
inner_fire_water.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1 entries, 0 to 0
Data columns (total 24 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   #             1 non-null      int64 
 1   Name          1 non-null      object
 2   Type_1_x      1 non-null      object
 3   Type_2_x      1 non-null      object
 4   Total_x       1 non-null      int64 
 5   HP_x          1 non-null      int64 
 6   Attack_x      1 non-null      int64 
 7   Defense_x     1 non-null      int64 
 8   Sp_Atk_x      1 non-null      int64 
 9   Sp_Def_x      1 non-null      int64 
 10  Speed_x       1 non-null      int64 
 11  Generation_x  1 non-null      int64 
 12  Legendary_x   1 non-null      bool  
 13  Type_1_y      1 non-null      object
 14  Type_2_y      1 non-null      object
 15  Total_y       1 non-null      int64 
 16  HP_y          1 non-null      int64 
 17  Attack_y      1 non-null      int64 
 18  Defense_y     1 non-null      int64 
 19  Sp_Atk_y    

Como podemos ver, os resultados são de fato bem diferentes.<br>
No *left*, mantemos os dados do dataframe à esquerda, e trazemos os dados do dataframe à direita no qual encontrou-se a chave (neste exemplo, o número e nome do Pokémon). Por outro lado, no *right* ocorre o contrário, mantemos os dados do dataframe à direita e, quando há correspondência da chave, trazemos os dados do dataframe à esquerda.
Por fim, o *inner* mantém apenas os dados dos Pokémons encontrados nos dois dataframes, dessa forma, a posição do dataframe não faz diferença.

Podemos também querer apenas concatenar dois *DataDrames*, isto é, juntá-los colocando um abaixo ou ao lado do outro. Para isso, utilizamos o método *.concat()*:

In [ ]:
fire_and_water = pd.concat([fire_pkmn, water_pkmn], ignore_index=True)
fire_and_water.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           190 non-null    int64 
 1   Name        190 non-null    object
 2   Type_1      190 non-null    object
 3   Type_2      103 non-null    object
 4   Total       190 non-null    int64 
 5   HP          190 non-null    int64 
 6   Attack      190 non-null    int64 
 7   Defense     190 non-null    int64 
 8   Sp_Atk      190 non-null    int64 
 9   Sp_Def      190 non-null    int64 
 10  Speed       190 non-null    int64 
 11  Generation  190 non-null    int64 
 12  Legendary   190 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 18.1+ KB


Acima fizemos a concatenação vertical. Vamos fazer a horizontal abaixo:

In [ ]:
atk_and_defense = pd.concat([offensive_stats, defensive_stats], axis=1)
atk_and_defense.head()

#                   Name  Attack  ...  HP  Defense  Sp_Def
0  1              Bulbasaur      49  ...  45       49      65
1  2                Ivysaur      62  ...  60       63      80
2  3               Venusaur      82  ...  80       83     100
3  3  VenusaurMega Venusaur     100  ...  80      123     120
4  4             Charmander      52  ...  39       43      50

[5 rows x 10 columns]

Você deve estar se perguntando: mas então qual a diferença entre utilizar o *merge* e o *concat* com axis=1 (concatenação horizontal)? Observe a imagem abaixo.
![concat](https://i.imgur.com/YlmiwsR.png) <br>

Note que o *concat* recebe os dataframes e apenas os empilha (verticalmente ou horizontalmente). Observe agora o funcionamento do *merge* na imagem abaixo.
![merge](https://i.imgur.com/yGum2id.png) <br>

Com o *merge*, podemos combinar os dataframes de acordo com os valores de suas colunas. Passamos a coluna a ser utilizada como chave, e os valores serão apenas combinados caso haja correspondência nos dois dataframes.

### 3.5. Alterando o dataframe

Até o momento apenas utilizamos os dados da forma que nos foram fornecidos, mas e se precisássemos criar alguma coluna que fosse a combinação das demais? Por exemplo, caso eu deseje criar uma coluna que corresponde à soma do ataque e velocidade dos Pokémons? Observe abaixo:

In [ ]:
#Criando a coluna desejada
pkmn['Sum_Attack_Speed'] = pkmn['Attack'] + pkmn['Speed']
pkmn.head()

#                   Name Type_1  ... Generation  Legendary  Sum_Attack_Speed
0  1              Bulbasaur  Grass  ...          1      False                94
1  2                Ivysaur  Grass  ...          1      False               122
2  3               Venusaur  Grass  ...          1      False               162
3  3  VenusaurMega Venusaur  Grass  ...          1      False               180
4  4             Charmander   Fire  ...          1      False               117

[5 rows x 14 columns]

Observe como foi fácil! Apenas utilizamos o operador de soma com as duas colunas necessárias. Você pode fazer isso com outras operações também, basta utilizar ```-```, ```/``` ou ```*```. Além disso, você pode combinar quantas colunas quiser!

Mas e se precisarmos alterar apenas algumas linhas do nosso DataFrame?

Por exemplo, suponha que você percebeu que seus dados estão errados, e todos os Pokémons com velocidade acima de 100 deveriam estar marcados como Type_1 = 'Fire', podemos seguir o procedimento abaixo:

In [ ]:
#Observe os valores unicos da coluna Type_1 para os Pokémons com mais de 100 de velocidade
pkmn.loc[pkmn['Speed']>100, 'Type_1'].unique()

array(['Bug', 'Normal', 'Electric', 'Ground', 'Psychic', 'Fire', 'Ghost',
       'Water', 'Rock', 'Poison', 'Grass', 'Dark', 'Dragon', 'Steel',
       'Fighting', 'Ice', 'Flying'], dtype=object)

In [ ]:
#Vamos alterar tudo para Fire
pkmn.loc[pkmn['Speed']>100, 'Type_1'] = 'Fire'

In [ ]:
#Observe como os valores mudaram
pkmn.loc[pkmn['Speed']>100, 'Type_1'].unique()

array(['Fire'], dtype=object)

In [ ]:
#Vamos alterar tudo para Fire
pkmn.loc[pkmn['Speed']>100, 'Type_1'] = 'Fire'

Antes de continuar, vamos ler novamente os dados de Pokémon, sem essa última alteração. Execute a célula abaixo.

In [ ]:
pkmn = pd.read_csv(
    'pokemon_data.csv', #o caminho para o arquivo que se quer ler
    sep=',') #o caracter utilizado para separar os valores

pkmn.rename(
    columns={'Type 1':'Type_1', 'Type 2':'Type_2', 'Sp. Atk':'Sp_Atk','Sp. Def':'Sp_Def'}, #passando o nome antigo e novo como um dicionário
    inplace = True #algumas operações com Pandas criam uma cópia do DataFrame e não alteram o objeto em si, alteramos isso mudando o parâmetro inplace para verdadeiro
)

### 3.6. Operações em grupo

Com Pandas nós podemos aplicar operações em grupos usando o método *.groupby()*. Ele é muito útil por ser uma forma bem simples de extrair informação de dados agregados. Para utilizá-lo, passamos as colunas nas quais queremos agrupar os dados e a operação que queremos fazer. Para exemplificar, vamos ver quantos Pokémons lendários cada geração tem:

In [ ]:
pkmn.groupby('Generation').Legendary.sum() #fazendo uma soma pois a coluna Legendary é boolean

Generation
1     6
2     5
3    18
4    13
5    15
6     8
Name: Legendary, dtype: int64

Podemos obter um relatório da média de diversas colunas para cada tipo de Pokémon:

In [ ]:
pkmn.groupby('Type_1')[['HP','Attack','Defense','Sp_Atk','Sp_Def','Speed']].mean().sort_values(by = "HP", ascending = False)

HP      Attack     Defense     Sp_Atk     Sp_Def       Speed
Type_1                                                                       
Dragon    83.312500  112.125000   86.375000  96.843750  88.843750   83.031250
Normal    77.275510   73.469388   59.846939  55.816327  63.724490   71.551020
Fairy     74.117647   61.529412   65.705882  78.529412  84.705882   48.588235
Ground    73.781250   95.750000   84.843750  56.468750  62.750000   63.906250
Water     72.062500   74.151786   72.946429  74.812500  70.517857   65.964286
Ice       72.000000   72.750000   71.416667  77.541667  76.291667   63.458333
Flying    70.750000   78.750000   66.250000  94.250000  72.500000  102.500000
Psychic   70.631579   71.456140   67.684211  98.403509  86.280702   81.491228
Fire      69.903846   84.769231   67.769231  88.980769  72.211538   74.442308
Fighting  69.851852   96.777778   65.925926  53.111111  64.703704   66.074074
Grass     67.271429   73.214286   70.800000  77.500000  70.428571   61.928571
Poison    67.250000   74.678571   68.821429  60.428571  64.392857   63.571429
Dark      66.806452   88.387097   70.225806  74.645161  69.516129   76.161290
Rock      65.363636   92.863636  100.795455  63.340909  75.477273   55.909091
Steel     65.222222   92.703704  126.370370  67.518519  80.629630   55.259259
Ghost     64.437500   73.781250   81.187500  79.343750  76.468750   64.343750
Electric  59.795455   69.090909   66.295455  90.022727  73.704545   84.500000
Bug       56.884058   70.971014   70.724638  53.869565  64.797101   61.681159

Isso é realmente muito importante e extremamente utilizado com pandas pois conseguimos fazer análises dos grupos com apenas uma linha de código. Podemos perceber, por exemplo, que Pokémons do tipo *Flying* são especialistas em velocidade enquanto *Dragon* e *Fighting* são especialistas em ataque.

### 3.7. Aplicando funções no Pandas

Com Pandas, nós temos um grande nível de controle de nossos dados, e somos capazes de transformá-los conforme precisarmos. Nós podemos, até mesmo, executar funções em DataFrames e manipulá-los como quisermos. Vamos revisitar o método head():

In [ ]:
pkmn.head()

#                   Name Type_1  ... Speed  Generation  Legendary
0  1              Bulbasaur  Grass  ...    45           1      False
1  2                Ivysaur  Grass  ...    60           1      False
2  3               Venusaur  Grass  ...    80           1      False
3  3  VenusaurMega Venusaur  Grass  ...    80           1      False
4  4             Charmander   Fire  ...    65           1      False

[5 rows x 13 columns]

Existem algumas mega evoluções misturadas no dataset (apenas alguns Pokémons são capazes de evoluir temporariamente para sua forma Mega, uma forma mais poderosa). Não seria legal se nós tivéssemos alguma flag que nos diria se um pokémon é mega ou não? E, por um acaso, será que os pokémons mega são mais poderosos?

Você deve ter percebido que evoluções mega têm um padrão em nosso DataFrame, algo como 'PokemonMega Pokemon'. Se nós tivermos esse padrão, podemos construir uma função que retorna True se este padrão for detectado:

In [ ]:
def is_it_mega(pokemon_name):
    """
    Recebe um nome de pokemon e diz se é uma mega evolução ou não
    I: string pokemon_name
    O: boolean para Mega evos
    """
    if 'Mega ' in pokemon_name: #é importante usar 'Mega ' e não 'mega', pois há um pokemon chamado called Yanmega e outro chamado Meganium que não são uma mega evolução
        return True
    else:
        return False

Vamos ver se funciona:

In [ ]:
is_it_mega('VenusaurMega Venusaur')

True

In [ ]:
is_it_mega('Squirtle')

False

Excelente! Seria ótimo se conseguíssemos aplicar essa função em todo nosso DataFrame. Para fazer isso, usaremos o método .apply(). Também criaremos uma coluna que é uma flag se o pokémon é mega:

In [ ]:
pkmn['Mega'] = pkmn.apply(
    lambda row: is_it_mega(row['Name']), #chamando uma função lambda que acabamos de construir
    axis=1 #qual direção queremos executar a função. 0 para horizontal, 1 para vertical
)

In [ ]:
pkmn.head()

#                   Name Type_1  Type_2  ...  Speed  Generation  Legendary   Mega
0  1              Bulbasaur  Grass  Poison  ...     45           1      False  False
1  2                Ivysaur  Grass  Poison  ...     60           1      False  False
2  3               Venusaur  Grass  Poison  ...     80           1      False  False
3  3  VenusaurMega Venusaur  Grass  Poison  ...     80           1      False   True
4  4             Charmander   Fire     NaN  ...     65           1      False  False

[5 rows x 14 columns]

Agora, vamos verificar quão poderosos são os pokémons mega:

In [ ]:
pkmn.groupby('Mega').Total.mean()

Mega
False    423.457447
True     617.541667
Name: Total, dtype: float64

Uau! Eles têm quase 200 stat points a mais que pokémons normais! Evoluções mega são, sim, muito poderosos! Uma boa prática é sempre tentar manter nosso DataFrame organizado. A forma como os pokémons mega estão nomeados não é muito ótima, e nós já temos uma coluna com a flag para pokémons Mega, então, vamos atacar isso! A estrutura do nome de um pokémon mega é da seguinte forma: 'NomeMega Nome'. Portanto, se nós pegarmos o qeu vem após o caractere ' ', teremos o nome original do pokémon!

In [ ]:
pkmn[(pkmn["Mega"] == True) & pkmn["Legendary"] == True]

#                   Name   Type_1  ... Generation  Legendary  Mega
163  150    MewtwoMega Mewtwo X  Psychic  ...          1       True  True
164  150    MewtwoMega Mewtwo Y  Psychic  ...          1       True  True
418  380      LatiasMega Latias   Dragon  ...          3       True  True
420  381      LatiosMega Latios   Dragon  ...          3       True  True
426  384  RayquazaMega Rayquaza   Dragon  ...          3       True  True
796  719    DiancieMega Diancie     Rock  ...          6       True  True

[6 rows x 14 columns]

In [ ]:
pkmn.Name.nunique() #conta elementos únicos de uma determinada coluna

800

In [ ]:
def get_original_name(s):
    """
    Recebe um nome de pokemon e retorna seu nome original
    I: s string
    O: string
    """
    return s.split(' ')

# pkmn['Name'] = pkmn.Name.apply(lambda s: get_original_name(s)) #sobreescrevendo a coluna Name
# pkmn.Name.nunique()

In [ ]:
pkmn.tail()

#                 Name   Type_1  ... Generation  Legendary   Mega
795  719              Diancie     Rock  ...          6       True  False
796  719  DiancieMega Diancie     Rock  ...          6       True   True
797  720  HoopaHoopa Confined  Psychic  ...          6       True  False
798  720   HoopaHoopa Unbound  Psychic  ...          6       True  False
799  721            Volcanion     Fire  ...          6       True  False

[5 rows x 14 columns]